From 67931365f1dea684eeda344f4e29bbc1b181a788 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 21 Aug 2025 16:44:54 +0800
Subject: [PATCH 235/236] mtk: hostapd: replace callback is_mld_finished() with
 get_finished_radio_mask()

The function is_mld_finished() only applies to MLD interfaces. In
order to make the Extender setup compatible with legacy interfaces,
it has been changed to use get_finished_radio_mask() to calculate
which AP radios have currently completed setup.

Also add get_radio_mask() to help ucode's STA startup request.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ucode.c | 42 ++++++++++++++++++++++++++----------------
 1 file changed, 26 insertions(+), 16 deletions(-)

diff --git a/src/ap/ucode.c b/src/ap/ucode.c
index 40cefcae4..0388ccfa8 100644
--- a/src/ap/ucode.c
+++ b/src/ap/ucode.c
@@ -721,26 +721,35 @@ uc_hostapd_iface_switch_channel(uc_vm_t *vm, size_t nargs)
 }
 
 static uc_value_t *
-uc_hostapd_iface_is_mld_finished(uc_vm_t *vm, size_t nargs)
+uc_hostapd_iface_get_radio_mask(uc_vm_t *vm, size_t nargs)
 {
 	struct hostapd_iface *iface = uc_fn_thisval("hostapd.iface");
-	bool finished = true;
-	int i;
 
-	for (i = 0; i < iface->num_bss; i++) {
-		struct hostapd_data *bss = iface->bss[i];
-
-		if (bss->conf->mld_ap && bss->mld) {
-			if (bss->conf->mld_allowed_links > 0 &&
-			    (1 << bss->mld->refcount) - 1 !=
-			    bss->conf->mld_allowed_links) {
-				finished = false;
-				break;
-			}
-		}
+	if (!iface)
+		return ucv_int64_new(0);
+
+	return ucv_int64_new(iface->radio_mask);
+}
+
+static uc_value_t *
+uc_hostapd_iface_get_finished_radio_mask(uc_vm_t *vm, size_t nargs)
+{
+	struct hostapd_iface *iface = uc_fn_thisval("hostapd.iface");
+	struct hapd_interfaces *interfaces;
+	int i, radio_mask = 0;
+
+	if (!iface)
+		return ucv_int64_new(0);
+
+	interfaces = iface->interfaces;
+	for (i = 0; i < interfaces->count; i++) {
+		struct hostapd_iface *tmp_iface = interfaces->iface[i];
+
+		if (tmp_iface->current_hw_info)
+			radio_mask |= BIT(tmp_iface->current_hw_info->hw_idx);
 	}
 
-	return ucv_boolean_new(finished);
+	return ucv_int64_new(radio_mask);
 }
 
 static uc_value_t *
@@ -920,7 +929,8 @@ int hostapd_ucode_init(struct hapd_interfaces *ifaces)
 		{ "stop", uc_hostapd_iface_stop },
 		{ "start", uc_hostapd_iface_start },
 		{ "switch_channel", uc_hostapd_iface_switch_channel },
-		{ "is_mld_finished", uc_hostapd_iface_is_mld_finished },
+		{ "get_radio_mask", uc_hostapd_iface_get_radio_mask },
+		{ "get_finished_radio_mask", uc_hostapd_iface_get_finished_radio_mask },
 	};
 	uc_value_t *data, *proto;
 
-- 
2.45.2

